pydantic/networks.py

```diff
@@ -489,13 +482,41 @@ def __get_pydantic_core_schema__(
             serialization=core_schema.to_string_ser_schema(),
         )
 
+        json_schema = core_schema.no_info_after_validator_function(
+            cls._validate,
+            core_schema.union_schema(
+                [
+                    core_schema.str_schema(metadata={'format': 'name-email'}),
+                    core_schema.typed_dict_schema(
+                        {
+                            'name': core_schema.typed_dict_field(core_schema.str_schema()),
+                            'email': core_schema.typed_dict_field(core_schema.str_schema()),
+                        }
+                    ),
+                ],
+                custom_error_type='name_email_type',
+                custom_error_message='Input is not a valid NameEmail',
+            ),
+            serialization=core_schema.to_string_ser_schema(),
+        )
+
+        return core_schema.json_or_python_schema(
+            python_schema=python_schema,
+            json_schema=json_schema,
+            serialization=core_schema.to_string_ser_schema(),
+        )
+
     @classmethod
-    def _validate(cls, __input_value: NameEmail | str) -> NameEmail:
+    def _validate(cls, __input_value: NameEmail | dict[str, str] | str) -> NameEmail:
```

Comment:
With the new core schema that I suggested above, I don't think you'll need to modify this function at all.

I don't think we want to support the `dict[str, str]` case here.